### PR TITLE
Add step option where='edges' to facilitate pre-binned hist plots

### PR DESCRIPTION
**Pull request by andrzejnovak**
## PR Summary

First attempt to address one of the requests raised in #6669 simplifying the api of step for pre-binned histogram plotting like 
```
data = np.random.normal(4,1, 200)
y, x = np.histogram(data, bins=np.linspace(0,8,9))
```
from 
```
fig, ax = plt.subplots()
ax.step(x,np.r_[y, y[-1]], where='post');
```
to
```
fig, ax = plt.subplots()
ax.step(x, y, where='edges');
```

The consensus seems to have landed on no lines at the end, but this could easily be switched.
image
In this specific case I am also not sure whether 'edges' is the best keyword, maybe 'bins' would be better if we don't have the end lines go to -inf.

I'll be happy about feedback on the actual implementation.

Also it would be good to know if it's better to address the various enhancements in #6669 one by one or to combine things into less PRs.

## PR Checklist

- [ ] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -2089,6 +2089,8 @@ def step(self, x, y, *args, where='pre', data=None, **kwargs):
               every *x* position, i.e. the interval ``[x[i], x[i+1])`` has the
               value ``y[i]``.
             - 'mid': Steps occur half-way between the *x* positions.
+            - 'edges': Expects dim(x) = dim(y) + 1, steps have y[i] value on
```
Comment (dstansby) [inline lib/matplotlib/axes/_axes.py]: I think `len(x)` instead of `dim(x)` here?

**Comment by timhoffm** [comment]
I assume, if the fist *y* is at 50, there would be no vertical line y=0 to y=50? E.g. something like (not tested with your code)

grafik

Is that intentional? I could imagine many people would want that (optionally or by default). Otherwise this is not similar to a histogram.

**Comment by timhoffm** [comment]
Closes #5855.

**Comment by andrzejnovak** [comment]
Yes, that's intentional based on the consensus in @6669. I would be ok with having both options, as we might want the above implementation to signify that we have no info about the rest of the distribution, but other might want the same results as with `ax.hist(data, bins=x, histtype='step')` just with preprocessed data.

Naively I would expect  `"edges"` to actually refer to the second option with closed lines, so it would make more sense to call the above `"bins"` or just descriptively `"between"`.

**Comment by anntzer** [comment]
As a reminder, there's also https://github.com/matplotlib/matplotlib/issues/5855 where I suggested allowing both len(x) == len(y)-1 (the case here) and len(y) == len(x)+1 (with "edges", or altenatively this can be seen as a histogram flipped by 90° -- indeed, my intent was to make it easy to switch between one orientation and the other).

**Comment by andrzejnovak** [comment]
Tbh not sure how to implement that, but also I don't really like the idea of implicitly flipping x and y.

**Comment by anntzer** [comment]
It does not have to be implicit, but let's say I would prefer it not to be *much* harder to do a horizontal histogram rather than a vertical one.

**Comment by andrzejnovak** [comment]
On that note, there seems to be some inconsistency on how `np.inf` is handled. 
```
fig, ax = plt.subplots()
ax.step(np.r_[x, x[-1]],[-np.inf,*y, -np.inf], where='pre');
ax.set_ylim(1,50)
ax.set_yscale('log')
```

and 
```
fig, ax = plt.subplots()
ax.step([*x, x[-1]],[-np.inf,*y, -np.inf], where='pre');
ax.set_ylim(1,50);
```
Either show or don't show the last edge.

**Comment by andrzejnovak** [comment]
The issues with having the closing edges is that the only way to make it work seems if one could specify transform point by point. Right now it looks like it's possible to have x in data coords and y in axes coords, but I don't see how to do it for an line with coords [data, data] [data, axes]. If there's no better solution I'll make it default to a small nonzero number.

**Comment by anntzer** [comment]
you can (should be able to... no guarantees) be able to make it zero and the log-scale handler will clip it itself to a small nonzero number.

**Comment by andrzejnovak** [comment]
```
fig, ax = plt.subplots()
ax.step(x, y, where='edges');
```
image

```
fig, ax = plt.subplots()
ax.step(x, y, where='edges')
ax.semilogy();
```
image

```
fig, ax = plt.subplots()
ax.step(x,y, where='between');
```
image

**Comment by andrzejnovak** [comment]
@anntzer I added a `flipxy` bool to do what you wanted in #5855, could you take a look if that implementation works for you?

**Comment by anntzer** [comment]
I guess a variant would be to have orientation="horizontal"/"vertical" which is consistent with the rest of matplotlib (whereas flipxy exists nowhere else).  Probably needs some discussion...

**Comment by andrzejnovak** [comment]
I had that at first, but apart from being a handful to write I got confused which way was ''vertical" and which way was "horizontal". I don't have a strong feeling either way. Would be great to get a consensus, so I can start putting in tests.

**Comment by anntzer** [comment]
The orientation would be the same as for hist().  I agree it's not clear this API would be best, just suggesting it.

**Comment by andrzejnovak** [comment]
@dstansby I've added tests, could you reveiw and tell me what's missing?

File: lib/matplotlib/tests/test_axes.py [hunk 2]
```diff
@@ -3696,6 +3696,28 @@ def test_step_linestyle():
         ax.set_ylim([-1, 7])
 
 
+@image_comparison(['step_histlike'], remove_text=True, extensions=['png'])
+def test_step_histlike():
```
Comment (anntzer) [inline lib/matplotlib/tests/test_axes.py]: consider writing the test using check_figures_equal, which would better test that using step(..., where="between") is indeed equivalent to some call to plot()

**Comment by anntzer** [comment]
In general I like the idea, but I won't have the time to properly review this in the coming days at least.  I'm not claiming exclusive rights to this PR either :)

**Comment by anntzer** [comment]
Although a quick thought is that after all, it may not be such a great idea to have an 'orientation' kwarg which *only* has an effect when 'where' has a certain value ("edges"), so perhaps instead having where="edges-horizontal"/"edges-vertical" (or some better name, but basically folding this into 'where') may be better.
Not really sure here, just some thoughts.

**Comment by andrzejnovak** [comment]
Well it works for both new symmetric where options. I didn't modify the rest of them because its not clear to me what the behavior of pre and post should look like.

Tbh I'd be curious to know what fraction of step calls is not plotting histograms.

**Comment by anntzer** [comment]
At least when I first proposed the idea in https://github.com/matplotlib/matplotlib/issues/5855 the goal was not histograms (see https://github.com/matplotlib/matplotlib/issues/5855#issuecomment-171962098 and https://github.com/matplotlib/matplotlib/issues/5855#issuecomment-172021134).

**Comment by andrzejnovak** [comment]
Apart from the obvious I fail to see what's different about [this comment](https://github.com/matplotlib/matplotlib/issues/5855#issuecomment-172021134) and as implemented here `where='between'` (or that but horizontal). Feel free to put up an image of what you mean and I can try to work it in.

As for [this comment](https://github.com/matplotlib/matplotlib/issues/5855#issuecomment-171962098) we're planning to make something at least for mpl-hep of the signature `histplot(values, bin_edges, histtype=<outline/filled/errors>) which will under the hood call `step()/fill_between()/errorbar()` or their combinations. If interested we'd be happy to work it into matplotlib itself. But this PR already should simplify our life quite a bit.

**Comment by ImportanceOfBeingErnest** [comment]
I left my problem with this implementation as [a comment to #5855](https://github.com/matplotlib/matplotlib/issues/5855#issuecomment-521839135).

File: lib/matplotlib/axes/_axes.py [hunk 3]
```diff
@@ -2104,7 +2117,37 @@ def step(self, x, y, *args, where='pre', data=None, **kwargs):
...
+        if where == 'between' or where == 'edges':
+            if x.shape[0] != y.shape[0] + 1:
+                raise ValueError(f"When drawing with 'edges' or 'between', "
+                                 f"x must have first dimension greater "
+                                 f"then y by exactly 1, but x, y "
+                                 f"have shapes {x.shape} and "
+                                 f"{y.shape}")
+
+            if orientation == 'horizontal':
+                y = np.r_[y[0], y]
```
Comment (jklymak) [inline lib/matplotlib/axes/_axes.py]: Please see #16142 :wink:

Comment (andrzejnovak) [reply lib/matplotlib/axes/_axes.py]: Alright., btw this PR is superseded by #15065, I should close it

**Comment by jklymak** [comment]
See  #15065

**Pull request closed without merging**